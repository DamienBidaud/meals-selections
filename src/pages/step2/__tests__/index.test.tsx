import { render, RenderResult, screen, waitFor } from "@testing-library/react";
import { useEffect } from "react";
import { atom, RecoilRoot, useRecoilValue } from "recoil";
import Step2 from "../index";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import * as routerFunction from "react-router";
import userEvent from "@testing-library/user-event";

//@ts-ignore
export const RecoilObserver = ({ node, onChange }) => {
  const value = useRecoilValue(node);
  useEffect(() => onChange(value), [onChange, value]);
  return null;
};

const mockNavigate = jest.fn();
const mockUseNavigate = jest.spyOn(routerFunction, "useNavigate");

const router = createBrowserRouter([
  {
    path: "/",
    element: <Step2 />,
  },
]);
const formState = atom({
  key: "form",
  default: {
    meal: "dinner",
    guests: 3,
    restaurant: "",
    selectedMeals: [
      {
        meal: "",
        serving: 1,
      },
    ],
  },
});

const onChange = jest.fn();
describe("Step 2", () => {
  let step2: RenderResult;
  beforeEach(() => {
    step2 = render(
      <RecoilRoot>
        <RouterProvider router={router} />
        <RecoilObserver node={formState} onChange={onChange} />
      </RecoilRoot>
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });
  it("Should match the snapshot", () => {
    const { asFragment } = step2;
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should not navagitate when no restaurant have been selected", () => {
    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    mockUseNavigate.mockReturnValue(mockNavigate);

    userEvent.click(screen.getByText("Next"));
    expect(mockNavigate).not.toHaveBeenCalled();
    expect(alertMock).toHaveBeenCalledTimes(1);
  });

  it("Should navigate when a restaurant have been selected", () => {
    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    mockUseNavigate.mockReturnValue(mockNavigate);
    userEvent.selectOptions(screen.getByRole("combobox"), "Mc Donalds");

    userEvent.click(screen.getByText("Next"));
    expect(onChange).toHaveBeenCalledWith({
      guests: 3,
      meal: "dinner",
      restaurant: "Mc Donalds",
      selectedMeals: [
        {
          meal: "",
          serving: 1,
        },
      ],
    });
    expect(mockNavigate).toHaveBeenCalledWith("/step3");
    expect(alertMock).toHaveBeenCalledTimes(0);
  });
});
