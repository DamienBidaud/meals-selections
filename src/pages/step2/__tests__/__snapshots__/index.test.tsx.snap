// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step 2 Should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe bwnVWm"
  >
    <h1>
      Step 2
    </h1>
    <div>
      <label
        for="restaurantList"
      >
        Select a restaurant
      </label>
      <select
        id="restaurantList"
        name="restaunt"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          ----
        </option>
        <option
          value="Mc Donalds"
        >
          Mc Donalds
        </option>
        <option
          value="Taco Bell"
        >
          Taco Bell
        </option>
        <option
          value="BBQ Hut"
        >
          BBQ Hut
        </option>
        <option
          value="Vege Deli"
        >
          Vege Deli
        </option>
        <option
          value="Pizzeria"
        >
          Pizzeria
        </option>
        <option
          value="Panda Express"
        >
          Panda Express
        </option>
        <option
          value="Olive Garden"
        >
          Olive Garden
        </option>
      </select>
    </div>
    <div>
      <button
        class="sc-dkrFOg DBrcQ"
      >
        Back
      </button>
      <button
        class="sc-dkrFOg DBrcQ"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
